File: client/old/src/main/java/net/treset/minecraftlauncher/util/string/PatternString.java
package net.treset.minecraftlauncher.util.string;

import java.util.Arrays;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class PatternString extends FormatString {
    private String pattern;

    public PatternString(String original) {
        this(original, false);
    }

    public PatternString(String original, boolean keep) {
        if(keep) {
            this.pattern = original;
        } else {
            this.pattern = original.replaceAll("\\\\", "\\\\\\\\").replaceAll("(?<=^|[^\\\\])\\.(?=[^*+?]|$)", "\\\\.");
            if(!pattern.startsWith(".*")) pattern = "^" + pattern;
            if(!pattern.endsWith(".*")) pattern += "$";
        }
    }

    public String firstGroup(String test) {
        if(test == null) return null;
        Matcher matcher = Pattern.compile(get()).matcher(test);
        if(matcher.find()) {
            return matcher.group(1);
        }
        return null;
    }

    public boolean matches(String test) {
        if(test == null) return false;
        return test.matches(get());
    }

    // \ -> \\; . -> \.; [.*;.+;.?] unchanged; no .* at start / end -> ^ / $
    public String get() {
        return pattern;
    }

    public static List<PatternString> toPattern(String... items) {
        return Arrays.stream(items).map(PatternString::new).toList();
    }

    public static String decode(String pattern) {
        return pattern.replaceAll("\\\\\\\\", "\\\\")
                .replaceAll("\\\\\\.", ".")
                .replaceAll("^\\^", "")
                .replaceAll("\\$$", "");
    }

    public static List<String> decode(String... patterns) {
        return Arrays.stream(patterns).map(PatternString::decode).toList();
    }

    public static boolean matchesAny(String test, List<PatternString> patterns) {
        if(patterns == null) return false;
        return patterns.stream().anyMatch(p -> p.matches(test));
    }
}

File: client/composeApp/src/commonMain/kotlin/net/treset/treelauncher/backend/util/string/PatternString.kt
package net.treset.treelauncher.backend.util.string

import java.util.*
import java.util.regex.Pattern

class PatternString(original: String, keep: Boolean = false) : FormatString() {
    private val pattern: String

    init {
        pattern =
            if (keep) {
                original
            } else {
                original.replace("\\\\".toRegex(), "\\\\\\\\")
                    .replace("(?<=^|[^\\\\])\\.(?=[^*+?]|$)".toRegex(), "\\\\.")
                    .let {
                        if (it.startsWith("^")) it else "^$it"
                    }.let {
                        if (it.endsWith("$")) it else "$it$"
                    }
            }
    }

    fun firstGroup(test: String?): String? {
        if (test == null) return null
        val matcher = Pattern.compile(get()).matcher(test)
        return if (matcher.find()) {
            matcher.group(1)
        } else null
    }

    fun matches(test: String?): Boolean {
        return test?.matches(get().toRegex()) ?: false
    }

    // \ -> \\; . -> \.; [.*;.+;.?] unchanged; no .* at start / end -> ^ / $
    override fun get(): String {
        return pattern
    }

    companion object {
        fun toPattern(vararg items: String): Array<PatternString> {
            return items.map { original: String -> PatternString(original) }.toTypedArray()
        }

        fun decode(pattern: String): String {
            return pattern.replace("\\\\\\\\".toRegex(), "\\\\")
                .replace("\\\\\\.".toRegex(), ".")
                .replace("^\\^".toRegex(), "")
                .replace("\\$$".toRegex(), "")
        }

        fun decode(vararg patterns: String): Array<String> {
            return patterns.map { pattern: String -> decode(pattern) }.toTypedArray()
        }

        fun matchesAny(test: String, patterns: Array<PatternString>): Boolean {
            return patterns.any { p: PatternString -> p.matches(test) }
        }
    }
}
